### PR TITLE
topology2: use a macro to config if a sdw amplifier support feedback

### PR DESCRIPTION
**Pull request by bardliao**
Not all amplifiers support feedback. We should not add feedback support
on topology to shch amplifiers.

### CODE REVIEW

File: tools/topology/topology2/platform/intel/sdw-amp-generic.conf
Comment (plbossart) [inline tools/topology/topology2/platform/intel/sdw-amp-generic.conf:60]: indentation?

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (plbossart) [inline tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: I know I asked this before, but I don't recall what the answer was

Can we really encode this '515' value which is related to ALH channels? this might depend on which link and ALH channel is used for the feedback, no? So if this is in an include file should this be a parameter?

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (plbossart) [inline tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:17]: can we remove references to passthrough, the feedback is likely to be split and provided to the smart-amp algorithm

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (plbossart) [inline tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:24]: comment is very cryptic, what are you trying to say here?

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (plbossart) [inline tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:62]: what happens if we want to split the feedback? This is fine for an initial start, but we do know this is going to have to be split between the dai and host parts

File: tools/topology/topology2/avs-tplg/CMakeLists.txt
Comment (plbossart) [inline tools/topology/topology2/avs-tplg/CMakeLists.txt:32]: maybe a question for @ranj063 : if we want to have a boolean, should we use true/false rather than 1/0 in the key definition?

Comment (ranj063) [reply tools/topology/topology2/avs-tplg/CMakeLists.txt:32]: @plbossart It really depends on what the IncludeByKey uses for testing. So if we use 1 here, we can only test for 1 when including the file conditionally like below. But I think using "true" makes more sense intuitively.
```
IncludeByKey.SDW_AMP_FEEDBACK {
"1" "platform/intel/sdw-amp-feedback.conf"
```

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: > I know I asked this before, but I don't recall what the answer was
> 
> Can we really encode this '515' value which is related to ALH channels? this might depend on which link and ALH channel is used for the feedback, no? So if this is in an include file should this be a parameter?

@plbossart The ALH index (`515` here) is not related to ALH channels. We get `alh_stream_id` from [sdw_params_stream](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda.c#L92) and set [copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(data->dai_data);](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-topology.c#L2330)
The only restriction of ALH index is that we can't have duplicated ALH index in a topology. In other words, we can have ALH.0, ALH.1, ALH.2, ALH.3, ALH.4 if we want.

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:24]: > comment is very cryptic, what are you trying to say here?

That was copy paste mistake. I will remove the comment.

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:62]: > what happens if we want to split the feedback? This is fine for an initial start, but we do know this is going to have to be split between the dai and host parts

@plbossart I am not sure if I get your question. The dai and host are in different pipelines now. If we want to add some widgets between dai and host, we can add the widgets in the pipeline and add routes for them.

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:17]: > can we remove references to passthrough, the feedback is likely to be split and provided to the smart-amp algorithm

@plbossart If there is a smart-amp widget, what we need to do is to add a route from Copier.ALH.31.1 to the smart-amp widget. Copier.ALH.31.1 itself can work as a demux

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (plbossart) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: oh, so if this is a convention that's totally arbitrary we should not use 515 which was from another era and tied to a specific link/PDI usage. Should we use the same numbers as for PCM devices then in our ALH. numbering? I mean the speaker could be ALH.2 and mic ALH.4.

Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: > oh, so if this is a convention that's totally arbitrary we should not use 515 which was from another era and tied to a specific link/PDI usage. Should we use the same numbers as for PCM devices then in our ALH. numbering? I mean the speaker could be ALH.2 and mic ALH.4.

I choose the same ALH. number as for IPC3 ALH number is because that way can guarantee we will not use duplicated ALH. number. A problem of using the same number as for PCM devices is that we need two ALH in an aggregated speaker device. For example, currently we have ALH.258 and ALH.514 for one aggregated speaker device, and we can't use ALH.2-1 and ALH.2-2

Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: > the ALH number is not fixed for IPC3, it's based on macros. I don't mind if we use the same convention but we don't want to encode the link position in the topology, because we'll have to redo this when we support different hardware layouts.

@plbossart Yeah, I know the ALH number is not fixed for IPC3. I meant in the most common case, we have

SDW0 DP2 for headset playback
SDW0 DP3 for headset capture
SDW1 DP2 for 1st speaker playback
SDW2 DP2 for 2nd speaker playback
SDW3 DP3 for dmic capture

Based on above layout I assigned IPC4 ALH number with the same formula as IPC3. i.e. 0x002 for headset playback, 0x003 for headset capture, 0x102 for 1st speaker, etc.

But we don't really care about the real HW layouts. We can still use ALH.2 for headset playback if the headset codec is not on SDW0.
In other words, the ALH index number doesn't need to be changed when the HW layouts are changed.
And we can use the same convention to add a new ALH dai.

Does it make sense?

Comment (plbossart) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: sorry, I am not following what IPC3 convention you are referring to @bardliao since we don't have one? 

Or maybe you are referring to the initial CML designs with 1 and 2 amps?

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (RanderWang) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:60]: you can add some macro definition like linking HOST_PIPELINE_MODULE to DAI_PIPELINE_MODULE

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:2]: > sorry, I am not following what IPC3 convention you are referring to @bardliao since we don't have one?
> 
> Or maybe you are referring to the initial CML designs with 1 and 2 amps?

@plbossart I was referring to the (link_id * 256 + data port) formula. But I just realize that we can use the same rule as the pipeline id and the route id. I.e. pcm id * 10 ~ pcm id * 10 + 9.
I will update the PR to follow the rule.

**Review by lgirdwood — APPROVED** [review]
topology syntax looks fine, but needs @plbossart approval for sdw.

File: tools/topology/topology2/cavs-sdw.conf
Comment (plbossart) [inline tools/topology/topology2/cavs-sdw.conf:98]: I can't follow what the formula is in the commit message. there's a stray '~' character in there. Can we fix?

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (plbossart) [inline tools/topology/topology2/platform/intel/sdw-amp-feedback.conf:10]: is the 515 here intentional? I thought these were removed

**Review by plbossart — CHANGES_REQUESTED** [review]
This looks mostly good, just two small opens that should be trivial to fix or confirm as correct. Thanks @bardliao

File: tools/topology/topology2/platform/intel/sdw-amp-feedback.conf
Comment (bardliao) [reply tools/topology/topology2/platform/intel/sdw-amp-feedback.conf]: > is the 515 here intentional? I thought these were removed

No. I forgot to change the index in hw_config. However, somehow it still works.

File: tools/topology/topology2/cavs-sdw.conf
Comment (bardliao) [reply tools/topology/topology2/cavs-sdw.conf:98]: > I can't follow what the formula is in the commit message. there's a stray '~' character in there. Can we fix?

Fixed

**Review by RanderWang — APPROVED** [review]

File: tools/topology/topology2/cavs-sdw-src-gain-mixin.conf
Comment (plbossart) [inline tools/topology/topology2/cavs-sdw-src-gain-mixin.conf:42]: not completely following the formula. 

a) what is the relationship between pcm_id and ALH? there could be M:1 and 1:N topologies.
b) what is a route_id
c) why does the pipeline come into play.

That was a long day, bear with me

Comment (bardliao) [reply tools/topology/topology2/cavs-sdw-src-gain-mixin.conf:42]: > not completely following the formula.

@plbossart Please see https://github.com/thesofproject/sof/pull/6549/commits/f554a2f1714bec1c3b8b76844faebba8e5788096 for your reference. I am trying to follow the same customary to assign ALH index. 

> 
> a) what is the relationship between pcm_id and ALH? there could be M:1 and 1:N topologies. 

ALH index = pcm_id * 10 + N, where N could be 0 to 9.
For example, say there is an aggregated speaker device
pcm_id = 2, and the ALH index will be 20, 21.

> b) what is a route_id 

For example,
route.**0** {
		source	"gain.1.1"
		sink	"copier.ALH.1.1"
	}
The route_id above is 0

> c) why does the pipeline come into play.

The pipeline id also follows the same rule.


> 
> That was a long day, bear with me

Comment (bardliao) [reply tools/topology/topology2/cavs-sdw-src-gain-mixin.conf:42]: @plbossart Are you good with this PR?

**Review by lgirdwood — APPROVED** [review]

**Comment by ranj063** [comment]
@bardliao @lgirdwood Can you please wait before merging this PR? I have a more elegant solution for avoid clashes between included file objects and top-level objects.

**Comment by ranj063** [comment]
> @bardliao @lgirdwood Can you please wait before merging this PR? I have a more elegant solution for avoid clashes between included file objects and top-level objects.

@bardliao this is the change I was talking about. Could you please update your PR based on that one?

**Comment by bardliao** [comment]
> > @bardliao @lgirdwood Can you please wait before merging this PR? I have a more elegant solution for avoid clashes between included file objects and top-level objects.
> 
> @bardliao this is the change I was talking about. Could you please update your PR based on that one? #7113

@ranj063 I updated my PR based on #7113, will push after #7113 is merged.

**Comment by kv2019i** [comment]
@bardliao array-support merged in #7113 so rebase will be needed

**Comment by bardliao** [comment]
> @bardliao array-support merged in #7113 so rebase will be needed

done

File: tools/topology/topology2/cavs-sdw-src-gain-mixin.conf
Comment (ranj063) [inline tools/topology/topology2/cavs-sdw-src-gain-mixin.conf:42]: the commit message talks about pipeline ID and route ID but all im seeing is dai_index updates.

Comment (bardliao) [reply tools/topology/topology2/cavs-sdw-src-gain-mixin.conf:42]: > the commit message talks about pipeline ID and route ID but all im seeing is dai_index updates.

@ranj063 The pipeline ID and route ID is already grouped by f0a010052bd0. This commit follows the same rule for ALH dai index. I will update the commit message to make it more intuition.

**Review by lgirdwood — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Pull request merged**
